README.md

```diff
@@ -107,12 +113,97 @@ start locations.  More details on what is contained in the metadata can be found
 in the Thrift definition.
 
 Metadata is written after the data to allow for single pass writing.
+This is especially useful when writing to backends such as S3.
 
 Readers are expected to first read the file metadata to find all the column
 chunks they are interested in.  The columns chunks should then be read sequentially.
 
  ![File Layout](https://raw.github.com/apache/parquet-format/master/doc/images/FileLayout.gif)
 
+### Parquet 3
+
+Parquet 3 files have the following overall structure:
+
+```
+4-byte magic number "PAR1"
+4-byte magic number "PAR3"
+
+<Column 1 Chunk 1 + Column Metadata>
+<Column 2 Chunk 1 + Column Metadata>
+...
+<Column N Chunk 1 + Column Metadata>
+<Column 1 Chunk 2 + Column Metadata>
+<Column 2 Chunk 2 + Column Metadata>
+...
+<Column N Chunk 2 + Column Metadata>
+...
+<Column 1 Chunk M + Column Metadata>
+<Column 2 Chunk M + Column Metadata>
+...
+<Column N Chunk M + Column Metadata>
+
+<File-level Column 1 Metadata v3>
+...
+<File-level Column N Metadata v3>
+
+File Metadata v3
+4-byte length in bytes of File Metadata v3 (little endian)
+4-byte magic number "PAR3"
+
+File Metadata
+4-byte length in bytes of File Metadata (little endian)
+4-byte magic number "PAR1"
+```
+
+Unlike the legacy File Metadata, the File Metadata v3 is designed to be light-weight
+to decode, regardless of the number of columns in the file. Individual column
+metadata can be opportunistically decoded depending on actual needs.
+
+This file structure is backwards-compatible. Parquet 1 readers will read and
+decode the legacy File Metadata in the file footer, while Parquet 3 readers
+will notice the "PAR3" magic number just before the File Metadata and will
+instead read and decode the File Metadata v3.
```

Comment:
Can you suggest a layout that preserves compatibility with PAR1 readers?

Reply:
There are a few options:
1. Use `FileMetadata.version` to introduce a new version of the metadata. Starting from the minimal change that can be done in place (https://github.com/apache/parquet-format/pull/248) we can bump the version and remove `columns` from `RowGroup` and decouple the column metadata completely.
2. Add a binary field to `FileMetadata` named `v3_metadata` with tag number 10003. This field will encode the flatbuffer/thrift representation of the new footer. This field is going to be encoded last by thrift. Readers can manually look at the tail of the file and if they find this field, they can ignore the rest of the footer and parse these bytes only, ignoring the old style footer alltogether.

Reply:
The goal here is twofold:
1. achieve better metadata parsing performance for PAR3 readers
2. keep compatibility with PAR1 readers

This is why this proposal creates a separate array of structure types: so that PAR3 readers don't have to eagerly decode those pesky columns, while letting PAR1 readers correctly access column information.

I don't think any of your two proposals is able of achieving of two goals simultaneously, are they?
(admittedly, I'm not sure I understand proposal number 2, though it seems to require hand-coded Thrift parsing which doesn't sound like a tremendous idea)

Reply:
I think for 2, thrift could avoid parsing it assuming that we still follow the pattern of nested footer.

e.g. <field_marker 10003 and byte size><[serialized v3  metadatadata] + "PAR3"><footer size>PAR1

So the encoding/serialization would be manual but on decoding old readers should automatically drop the unknown field (it is possible some thrift implementations retain unknown fields, I know proto does) (i.e. the field ID 10003 should never actually be modeled in the schema).

So the trade-offs of doing this approach are:
1.  A bit of extra data to be copied for readers accessing the original version.
2. A guaranteed lower bound on amount of IO operations for V3 since it is incorporated into v2
3. Potentially more memory utilization if accessing the original version if unknown fields are maintained by thrift implementation.

Effectively for doing the operation currently as proposed in V3 the trade-offs are reverse.  I

Reply:
This sounds rather ad hoc and brittle to me.

Reply:
> This sounds rather ad hoc and brittle to me.

I would love to understand why. Preferably with an example.

Reply:
I'm not interested in spending hours convincing you that manual parsing of Thrift fields in a heroic attempt to remove overhead is a good idea. Let's agree to disagree.

Reply:
@pitrou curious if https://github.com/apache/parquet-format/pull/242/files#r1608953299 also sounds too brittle to you (it should not require any manual parsing of thrift fields, only some manual writes).

Reply:
It certainly does.

This is all in the name of an efficiency concern that doesn't seem that well-grounded. If your primary concern is to minimize the number of IO, you can also optimistically read, say, the last 4 MiB of the file and  iterate from that. Or the heuristic can be "length in bytes of File Metadata + 1 MiB".

Reply:
The overhead of putting the new footer before the old is substantial especially for the wide column use case. Footers for those files are in the order of MBs and parsing them is on the critical path of reads of any rowgroup/colchunk. The design as is will end up with 2-3 round trips to object stores for all the cases we care to optimise. Those fetches are going to take more time than parsing the old footer, effectively eliminating any benefits of the optimisations to the metadata.

`PAR3` are 4 arbitrary bytes as much as `{0x08, 0x93, 0xCE, 0x00}`. Except the latter have one additional property: they are are guaranteed to be ignored by thrift parsers along with what follows thrm. If you turn off the "oh my we are parsing thrift manually" alarm, this is not any more bad or scary than `PAR3`.

...

There might be another way. We could put the new footer after the old. Thrift terminates a struct with a null byte so it *knows* where the end is. What I don't know is if the thrift parser will fail at parsing thrift with trailing bytes after the end of a struct. I might have time to try it tomorrow.

Reply:
> you can also optimistically read, say, the last 4 MiB of the file and iterate from that

All good engines fetch the tail of the file to speculatively fetch both the size and the footer in a single fetch. A single fetch is important because fetching the footer is on the critical path. I am not aware of an engine implementation that does 4MB tail reads from object stores for parquet.

512KB S3 fetch is ~20ms.
2MB S3 fetch is ~30ms
4MB S3 fetch is ~50ms.
8MB S3 fetch is ~100ms.

A good engine does a 256-512kb tail fetch because most footers are small and reducing fetch size does not change the latency to get the bytes.

For reference parsing a 4mb footer (wide table) tables ~40ms.

If an engine leaves the tail fetch as is to avoid pessimizing reads for narrow tables, it means for wide tables it will have to do two fetches. Assuming new footer is typically 2x smaller than old:
- old: 20ms 256kb fetch + 50ms 4mb fetch + 40ms parse --> 110ms
- new before old: 20ms 256kb fetch + 80ms 6mb fetch + 2ms parse --> 102ms (1.1x speedup)
- new after old: 20ms 256kb fetch + 30ms 2mb fetch + 2ms parse --> 52ms (2.2x speedup)

If an engine increases the tail fetch, it will pessimize reads for narrow tables, with little if any benefit for wide tables.

From the above analysis and assuming we want to make parquet efficient for datalakes (parquet is stored in s3 like storage), having the new footer after the old is a *requirement*.

Reply:
With fetch latencies of that order, does the decode latency of the existing thrift payload even matter?

Reply:
Thanks for the concrete numbers. This gives us more precise insight for us to work on (though not all Parquet files are read from S3; local reads and writes are a very common use case as well). However, I'm a bit surprised by your numbers because typical guidance for S3 involves larger reads (i.e. multi-megabyte).

Reply:
Guidance from S3 involves larger reads are because of cost: reads from S3 typically cost an API call and bandwidth/transfer volume is free (same zone). 8MB reads will cost half the price of 4MB reads. Each connection can go up to ~100MB/sec so just transfering 4MB of data is at least 25ms. If one is reading large files, doing 100s of 8MB reads in parallel will saturate any network card - which is typically what engines do.

I posit that the vast majority of parquet encoded data is in some cloud somewhere. Hence working well with object stores (high latency, immutable files) is a requirement for any change. This is also lesson 4 from [An Empirical Evaluation of Columnar Storage Formats](https://arxiv.org/pdf/2304.05028).

Reply:
I posit that many PiB of parquet data lives in HDFS which also has read latencies, but not as bad as S3.
S3 standard has pretty bad latency for small byte reads plus you get billed. best to grab large amounts speculatively

Reply:
I'm late to this thread and haven't fully read it. I just want to give my +1 that the current proposal would require too many I/O requests and thus make this basically a deal breaker for high latency storage like S3. We would not use this due to this.

Any change that increases the number of data-dependent requests necessary to decode it basically a deal breaker for us and I guess is so for a lot of other data lake companies.